src/core/solvers/ipopt/multiple-shooting-nlp.cpp

```diff
@@ -0,0 +1,589 @@
+                                             bool           init_x,
+                                             Ipopt::Number *x,
+                                             bool           init_z,
+                                             Ipopt::Number *z_L,
+                                             Ipopt::Number *z_U,
+                                             Ipopt::Index   m,
+                                             bool           init_lambda,
+                                             Ipopt::Number *lambda)
+{
+    assert(init_x == true);
+    assert(init_z == false);
+    assert(init_lambda == false);
+
+    // initialize to the given starting point
+    // State variable are always at 0 since they represent increments from the given initial point
+    for (size_t i = 0; i < T_; i++) {
+        for (size_t j = 0; j < ndx_; j++) {
+            x[i * (ndx_ + nu_) + j] = 0;
+        }
+
+        for (size_t j = 0; j < nu_; j++) {
+            x[i * (ndx_ + nu_) + ndx_ + j] = us_[i](j);
+        }
+    }
+
+    for (size_t j = 0; j < ndx_; j++) {
+        x[T_ * (ndx_ + nu_) + j] = 0;
+    }
+
+    return true;
+}
+// [TNLP_get_starting_point]
+
+// [TNLP_eval_f]
+// returns the value of the objective function
+bool MultipleShootingNlp::eval_f(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Number &obj_value)
+{
+    assert(n == nvar_);
+
+    // Running costs
+    for (size_t i = 0; i < T_; i++) {
+        data_.dx      = Eigen::VectorXd::Map(x + i * (ndx_ + nu_), ndx_);
+        data_.control = Eigen::VectorXd::Map(x + i * (ndx_ + nu_) + ndx_, nu_);
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        problem_->get_runningModels()[i]->calc(problem_->get_runningDatas()[i], data_.x, data_.control);
+
+        obj_value += problem_->get_runningDatas()[i]->cost;
+    }
+
+    // Terminal costs
+    data_.dx = Eigen::VectorXd::Map(x + T_ * (ndx_ + nu_), ndx_);
+
+    state_->integrate(xs_[T_], data_.dx, data_.x);
+
+    problem_->get_terminalModel()->calc(problem_->get_terminalData(), data_.x);
+    obj_value += problem_->get_terminalData()->cost;
+
+    return true;
+}
+// [TNLP_eval_f]
+
+// [TNLP_eval_grad_f]
+// return the gradient of the objective function grad_{x} f(x)
+bool MultipleShootingNlp::eval_grad_f(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Number *grad_f)
+{
+    assert(n == nvar_);
+
+    for (size_t i = 0; i < T_; i++) {
+        data_.dx      = Eigen::VectorXd::Map(x + i * (ndx_ + nu_), ndx_);
+        data_.control = Eigen::VectorXd::Map(x + i * (ndx_ + nu_) + ndx_, nu_);
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        state_->Jintegrate(xs_[i], data_.dx, data_.Jsum_x, data_.Jsum_dx, crocoddyl::second, crocoddyl::setto);
+        problem_->get_runningModels()[i]->calcDiff(problem_->get_runningDatas()[i], data_.x, data_.control);
+        data_.Ldx = data_.Jsum_dx.transpose() * problem_->get_runningDatas()[i]->Lx;
```

Comment:
If I understand correctly this, then it would be better (more efficient) to use `JintegrateTransport`.
See this example: https://github.com/loco-3d/crocoddyl/blob/master/include/crocoddyl/core/integrator/euler.hxx#L113-L115